src/Property.hs

```diff
@@ -25,14 +27,14 @@ data PropertyResult =
   | Error String
   deriving (Eq, Show)
 
-runProperty :: Interpreter -> Expression -> IO PropertyResult
+runProperty :: Interpreter -> Expression -> Report PropertyResult
```

Comment:
You changed the type from `IO` to `Report`, however, I don't see where any `Report` specific functionality is used.

If this is not required, then I prefer you leave it as `IO` to minimize the diff.  That way reviewing will be faster.

If we want to do additional refactoring, we should do separately so that it's easy to review.

Reply:
There are two reasons for this change, a theoretical and a practical.

1. (Theoretical.) Running properties is, in principle, the same as running examples, just with another program. (With QuickCheck, as opposed to Hspec.) So, they should be designed in as similar a fashion as possible. For that, there needs to be a possibly similar set of instruments available. In particular, if the instruments that `Result` provides are available to example runner, they should be available for property runner as well.
2. (Practical.) I needed to use reporting functionality, and for that I needed to align the types. If I did not align the types here, I would have had to align them elsewhere, making more complicated, widespread, and less easily verifiable changes.

If you want me to split this pull request in two, this is unfortunate and you really should have made yourself available to discuss it earlier. And standing by itself, such refactoring is not quite as motivated.